include/solvers/petsc_nonlinear_solver.h

```diff
@@ -261,6 +262,16 @@ class PetscNonlinearSolver : public NonlinearSolver<T>
    */
   bool _computing_base_vector;
 
+  /**
+    *  Whether we have a live snes instance
+    */
+  bool _snes_allocated;
```

Comment:
`  WrappedPetsc<SNES> _snes;`  should have  enough information to detect whether or not SNES is created or not.  For example,

```
  /**
   * User-defined conversion to bool. This is intended to mimic code like:
   * IS is = nullptr;
   * ...
   * if (!is)
   *   ...
   * Note that this comparison is thus concerned with obj itself and
   * not &obj.
   */
  operator bool() const { return obj != nullptr; }
```
Considering that, I think should not need this extra flag `_snes_allocated`

Reply:
Yep, this is a good idea.